### PR TITLE
Making n-body benchmark work with LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
https://github.com/lfortran/lfortran/issues/1210

### CODE REVIEW

File: integration_tests/nbody.f90 [hunk 1]
```diff
@@ -0,0 +1,180 @@
+program nbody
```
Comment (certik) [inline integration_tests/nbody.f90:1]: Add a link where this was taken from, and license (if available), otherwise say copyright the original authors.

**Comment by czgdp1807** [comment]
@certik So this PR produces the following output,

```bash
(lf) 16:08:23:~/lfortran_project/lfortran % lfortran integration_tests/nbody.f90 -o nbody
(lf) 16:08:26:~/lfortran_project/lfortran % ./nbody
-1.69075163828524472e-01
-1.69087605234606420e-01
```

**Expected Output** (as per https://benchmarksgame-team.pages.debian.net/benchmarksgame/download/nbody-output.txt)

```bash
-0.169075164
-0.169087605
```

**Next Steps**

Fix `--fast` mode of LFortran for this benchmark. As of now the following happens,

```bash
(lf) 16:13:59:~/lfortran_project/lfortran % lfortran integration_tests/nbody.f90 --fast -o nbody
(lf) 16:14:56:~/lfortran_project/lfortran % ./nbody
                    nan
                    nan
```

@certik Do you want me to do this here or after merging this PR?

**Comment by certik** [comment]
Amazing! I would recommend:

* Add a test for the floating point result into the code, and do it for a smaller number of iterations, so that it runs quickly. If it agrees to all digits, it will work for the actual benchmark N as well.
* Let's merge this and run at our CI as a regular test.

Then let's identify the bugs in `--fast` and add an individual minimal test for each bug and fix them one by one.

File: integration_tests/nbody.f90 [hunk 2]
```diff
@@ -0,0 +1,200 @@
+! The Computer Language Benchmarks Game
+! https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
```
Comment (certik) [inline integration_tests/nbody.f90:2]: Add the exact link:

https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-ifc-6.html

**Comment by certik** [comment]
Looks like there are 2 workarounds that we currently had to implement in the code. How hard would it be to implement the missing features so that we can compile the code exactly as is?

**Comment by czgdp1807** [comment]
> Looks like there are 2 workarounds that we currently had to implement in the code

Can you please point out exactly which ones? Are you referring to the ones in https://github.com/lfortran/lfortran/issues/1210#issue-1553404946?

**Comment by certik** [comment]
I did some benchmarking on my machine. This C++ implementation: https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-gpp-9.html runs on my Apple M1 Max:
```console
lfortran/integration_tests(nbody) $ clang++ -Ofast -std=c++17  y.cpp 
lfortran/integration_tests(nbody) $ time ./a.out 50000000           
-0.169075164
-0.169059907
./a.out 50000000  1.93s user 0.01s system 99% cpu 1.937 total
```
The code in this PR runs at:
```console
lfortran/integration_tests(nbody) $ gfortran -Ofast nbody.f90 
lfortran/integration_tests(nbody) $ time ./a.out
 -0.16907516382852444     
 -0.16905990680106403     
ERROR STOP 

Error termination. Backtrace:
#0  0x102babc07
#1  0x102bac7a7
#2  0x102bada9b
#3  0x1027f3c9b
#4  0x1027f3cbf
./a.out  3.16s user 0.02s system 99% cpu 3.180 total
```
and
```console
lfortran/integration_tests(nbody) $ lfortran --fast nbody.f90 -o a.out
lfortran/integration_tests(nbody) $ time ./a.out                      
-1.69075163828524472e-01
-1.69059906784801178e-01
ERROR STOP
./a.out  7.95s user 0.01s system 99% cpu 7.969 total
```

So the `--fast` seems to work for me. There are a couple issues to look at:
* the number agrees with gfortran to single precision but not double precision
* we need to match gfortran performance
* then we need to match the C++ Clang performance, we can look at the LLVM code that it generates and match it

**Comment by certik** [comment]
We can get a little bit closer with this change:
```diff
--- a/integration_tests/nbody.f90
+++ b/integration_tests/nbody.f90
@@ -94,7 +94,7 @@ program nbody
 
     ! call getarg(1, argv)
     ! read (argv,*) num
-    num = 1000
+    num = 50000000
 
     call offsetMomentum(1, v, mass)
     e = energy(x, v, mass)
@@ -145,10 +145,8 @@ program nbody
         end do
 
         do m = 1, N
-            d2 = sum(r(:,m)**2)
-            distance = 1/sqrt(real(d2))
-            distance = distance * (1.5d0 - 0.5d0 * d2 * distance * distance)
-            !distance = distance * (1.5d0 - 0.5d0 * d2 * distance * distance)
+            d2 = r(1,m)**2 + r(2,m)**2 + r(3,m)**2
+            distance = 1/sqrt(d2)
             mag(m) = tstep * distance**3
         end do
 ```
Now it gives:
```console
lfortran/integration_tests(nbody) $ lfortran --fast nbody.f90 -o a.out
lfortran/integration_tests(nbody) $ time ./a.out                      
-1.69075163828524472e-01
-1.69059906813393585e-01
ERROR STOP
./a.out  6.91s user 0.03s system 99% cpu 6.950 total
```

**Comment by certik** [comment]
I was hoping you would create clean commits with the changes we had to make to the n-body. But since it's already merged, let it be.

**Comment by czgdp1807** [comment]
I did. See the last commit, https://github.com/lfortran/lfortran/pull/1213/commits/f2862a1a0d545b61eceb8f570cc86150c6081d6d. Did you want me to add detailed information of what the workaround actually is? If so then I create a new issue with the above commit and specify what the workaround is and what needs to done.

**Comment by czgdp1807** [comment]
Done, see, https://github.com/lfortran/lfortran/issues/1313

**Comment by certik** [comment]
It's good, thank you!

**Pull request merged**